labrad/decorators.py

```diff
@@ -21,202 +21,167 @@
-from twisted.internet.defer import inlineCallbacks
+import twisted.internet.defer as defer
 
 from labrad import types as T, util
 
-def _isGenerator(f):
-    """Check to see whether f is a generator.
-
-    See the documentation on code objects at:
-    http://docs.python.org/ref/types.html
-    """
-    return bool(f.func_code.co_flags & 0x20)
-
 def _product(lists):
     """Return the cartesian product of a list of lists."""
     if not len(lists): return [[]]
     return [[h] + t for h in lists[0] for t in _product(lists[1:])]
 
-class Setting(object):
-    def __init__(self, func):
-        self.func = func
+def setting(lr_ID, lr_name=None, returns=[], unflatten=True, **params):
+    """Decorator to turn a class method into a remotely accessible setting.
 
-    def getRegistrationInfo(self):
-        return (long(self.ID), self.name, self.description,
-                self.accepts, self.returns, self.notes)
-
-    def handleRequest(self, c, data):
-        return self.func(c, data)
-
-def setting(lr_ID, lr_name=None, returns=[], lr_num_params=2, **params):
-    """Mark a server method as a remotely-accessible setting.
-
-    The only required parameter is an integer ID.  You may
-    also provide a name to override the name of the decorated
-    function.  In addition, accepted types for each of the setting
-    parameters may be provided as named parameters with a list of
-    strings of allowed types.
+    This just creates a Setting class object which does the actual business.
     """
-    def decorated(f):
-        args, varargs, varkw, defaults = getargspec(f)
-        args = args[lr_num_params:]
+    def decorator(func):
+        return Setting(func, lr_ID, lr_name, returns, unflatten, **params)
+    return decorator
 
-        # handle generators as inlineCallbacks
-        if _isGenerator(f):
-            f = inlineCallbacks(f)
-
-        # make sure that defined params are actually accepted by the function.
-        # having extra params would not affect the running, but it is
-        # unnecessary and hence may indicate other problems with the code
-        for p in params:
+class Setting(object):
+    """
+    Represent a single setting within a Server object.
+
+    Args:
+
+    func (callable):   Function to implement the setting.
+    lr_ID (int):       The ID number of the setting used in the labrad protocol
+    lr_name (str):     The setting name.  By default, this is derived from the
+                       function name.
+    returns:           The labrad type tag or list of tags the setting returns
+    unflatten (bool):  Request automatic unflattening of incoming data.
+                       (default True)
+    **params:          Additional keyword arguments indicate arguments to the
+                       setting.  Each keyword should be a string matching one
+                       of the formal parameters of func, while the value is
+                       the type tag or list of tags accepted by that parameter.
```

Comment:
Should put the constructor arg docs on the constructor docstring instead of class docstring. Also, nit: args should be indented within the `Args:` section:
```
Args:
    func (callable): Function to implement the setting.
    ...
```